[quant][graphmode][fx][fp16] Add fp16 support for {add|mul}{_relu}

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #52865 [quant][graphmode][fx][fp16] Add fp16 support for silu
* #52864 [quant][graphmode][fx][fp16] Add fp16 support for tanh
* #52863 [quant][graphmode][fx][fp16] Add fp16 support for sigmoid
* #52862 [quant][graphmode][fx][fp16] Add fp16 support for layer_norm
* #52811 [quant][graphmode][fx][fp16] Add fp16 support for torch.sum
* #52810 [quant][graphmode][fx][fp16] Add fp16 support for div
* #52809 [quant][graphmode][fx][fp16] Add fp16 support for sub
* #52808 [quant][graphmode][fx] Add support for fp16 bmm pattern
* #52807 [quant][graphmode][fx][test][refactor] Refactoring binary op tests to split int8 and float16 tests
* **#52714 [quant][graphmode][fx][fp16] Add fp16 support for {add|mul}{_relu}**

Summary:

Test Plan:
python test/test_quantization.py TestQuantizeFxOps.test_add
python test/test_quantization.py TestQuantizeFxOps.test_mul
python test/test_quantization.py TestQuantizeFxOps.test_add_relu
python test/test_quantization.py TestQuantizeFxOps.test_mul_relu
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26626092](https://our.internmc.facebook.com/intern/diff/D26626092)

jerryzh168 committed:
[quant][graphmode][fx][fp16] Add fp16 support for {add|mul}{_relu}

Summary:

Test Plan:
python test/test_quantization.py TestQuantizedOps.test_add
python test/test_quantization.py TestQuantizedOps.test_mul
python test/test_quantization.py TestQuantizedOps.test_add_relu
python test/test_quantization.py TestQuantizedOps.test_mul_relu
Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx][fp16] Add fp16 support for {add|mul}{_relu}"


Summary:

Test Plan:
python test/test_quantization.py TestQuantizedOps.test_add
python test/test_quantization.py TestQuantizedOps.test_mul
python test/test_quantization.py TestQuantizedOps.test_add_relu
python test/test_quantization.py TestQuantizedOps.test_mul_relu
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26626092](https://our.internmc.facebook.com/intern/diff/D26626092)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx][fp16] Add fp16 support for {add|mul}{_relu}"


Summary:

Test Plan:
python test/test_quantization.py TestQuantizedOps.test_add
python test/test_quantization.py TestQuantizedOps.test_mul
python test/test_quantization.py TestQuantizedOps.test_add_relu
python test/test_quantization.py TestQuantizedOps.test_mul_relu
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26626092](https://our.internmc.facebook.com/intern/diff/D26626092)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx][fp16] Add fp16 support for {add|mul}{_relu}"


Summary:

Test Plan:
python test/test_quantization.py TestQuantizeFxOps.test_add
python test/test_quantization.py TestQuantizeFxOps.test_mul
python test/test_quantization.py TestQuantizeFxOps.test_add_relu
python test/test_quantization.py TestQuantizeFxOps.test_mul_relu
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26626092](https://our.internmc.facebook.com/intern/diff/D26626092)

[ghstack-poisoned]

vkuzo commented on torch/quantization/fx/quantization_patterns.py:
```diff
@@ -143,32 +144,44 @@ def convert(self, quantizer: QuantizerCls, node: Node, load_arg: Callable,
+                op = quantizer.quantized_graph.create_node(
+                    'call_function', self.qop, add_args, kwargs)
+                return op
+        elif dtypes in [(torch.float16, torch.float16, None)]:
```
just so I understand correctly...
1. this code copies the add or mul and relu
2. the code in the quantizer object will insert the fake quantize to fp16 logic, using `x.to(torch.half)`

?

jerryzh168 replied:
1. yes that's correct, in fp16 reference pattern we just copy add/mul op, the transformation we do happens in load_arg(quantized=False), which dequantizes the fp16 inputs.
2. not sure what you mean, the magic happens when people call quantize_node (which inserts x.to(torch.half)) and when we call load_arg(quantized=False), which inserts x = x.dequantize() (we can dequantize a float16 Tensor to float32 by calling to now: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/quantized/QTensor.cpp#L46)

jerryzh168 committed:
Update on "[quant][graphmode][fx][fp16] Add fp16 support for {add|mul}{_relu}"


Summary:

Test Plan:
python test/test_quantization.py TestQuantizeFxOps.test_add
python test/test_quantization.py TestQuantizeFxOps.test_mul
python test/test_quantization.py TestQuantizeFxOps.test_add_relu
python test/test_quantization.py TestQuantizeFxOps.test_mul_relu
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26626092](https://our.internmc.facebook.com/intern/diff/D26626092)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx][fp16] Add fp16 support for {add|mul}{_relu}"


Summary:

Test Plan:
python test/test_quantization.py TestQuantizeFxOps.test_add
python test/test_quantization.py TestQuantizeFxOps.test_mul
python test/test_quantization.py TestQuantizeFxOps.test_add_relu
python test/test_quantization.py TestQuantizeFxOps.test_mul_relu
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26626092](https://our.internmc.facebook.com/intern/diff/D26626092)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx][fp16] Add fp16 support for {add|mul}{_relu}"


Summary:

Test Plan:
python test/test_quantization.py TestQuantizeFxOps.test_add
python test/test_quantization.py TestQuantizeFxOps.test_mul
python test/test_quantization.py TestQuantizeFxOps.test_add_relu
python test/test_quantization.py TestQuantizeFxOps.test_mul_relu
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26626092](https://our.internmc.facebook.com/intern/diff/D26626092)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx][fp16] Add fp16 support for {add|mul}{_relu}"


Summary:

Test Plan:
python test/test_quantization.py TestQuantizeFxOps.test_add
python test/test_quantization.py TestQuantizeFxOps.test_mul
python test/test_quantization.py TestQuantizeFxOps.test_add_relu
python test/test_quantization.py TestQuantizeFxOps.test_mul_relu
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26626092](https://our.internmc.facebook.com/intern/diff/D26626092)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx][fp16] Add fp16 support for {add|mul}{_relu}"


Summary:

Test Plan:
python test/test_quantization.py TestQuantizeFxOps.test_add
python test/test_quantization.py TestQuantizeFxOps.test_mul
python test/test_quantization.py TestQuantizeFxOps.test_add_relu
python test/test_quantization.py TestQuantizeFxOps.test_mul_relu
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26626092](https://our.internmc.facebook.com/intern/diff/D26626092)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx][fp16] Add fp16 support for {add|mul}{_relu}"


Summary:

Test Plan:
python test/test_quantization.py TestQuantizeFxOps.test_add
python test/test_quantization.py TestQuantizeFxOps.test_mul
python test/test_quantization.py TestQuantizeFxOps.test_add_relu
python test/test_quantization.py TestQuantizeFxOps.test_mul_relu
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26626092](https://our.internmc.facebook.com/intern/diff/D26626092)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx][fp16] Add fp16 support for {add|mul}{_relu}"


Summary:

Test Plan:
python test/test_quantization.py TestQuantizeFxOps.test_add
python test/test_quantization.py TestQuantizeFxOps.test_mul
python test/test_quantization.py TestQuantizeFxOps.test_add_relu
python test/test_quantization.py TestQuantizeFxOps.test_mul_relu
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26626092](https://our.internmc.facebook.com/intern/diff/D26626092)

[ghstack-poisoned]

mruberry commented:
Unlanding this stack since it broke several HUD builds. Sample error:

```
test_run_mypy [mypy.ini] - TestTypeHints
test_type_hints.py
Traceback (most recent call last):
  File "test_type_hints.py", line 173, in test_run_mypy
    self.fail(f"mypy failed: {stdout} {stderr}")
AssertionError: mypy failed: torch/quantization/fx/quantization_patterns.py:116: error: Missing return statement  [return]
torch/quantization/fx/qconfig_utils.py:48: error: Missing return statement  [return]
Found 2 errors in 2 files (checked 1247 source files)
```
